yoyodyne/models/hmm.py

```diff
@@ -0,0 +1,413 @@
+        if self.enforce_monotonic:
+            alignment_matrix = self._apply_mono_mask(alignment_matrix)
+        return log_probs, alignment_matrix, decoder_hiddens
+
+    def greedy_decode(
+        self,
+        encoder_out: torch.Tensor,
+        encoder_mask: torch.Tensor,
+    ) -> torch.Tensor:
+        """Decodes a sequence given the encoded input.
+
+        Decodes until all sequences in a batch have reached [EOS] up to
+        a specified length depending on the `target` args.
+
+        Args:
+                encoder_out (torch.Tensor): batch of encoded input symbols.
+                    Shape: (batch_size x src_len x encoder_hidden*num_dir).
+                encoder_mask (torch.Tensor): mask for the batch of encoded
+                        input symbols. Shape: (batch_size x src_len).
+
+        Returns:
+                predictions (torch.Tensor): tensor of predictions of shape
+                        (batch_size x pred_seq_len).
+                likelihood (torch.Tensor): tensor of final likelihoods per
+                    prediction step. Shape (batch_size x 1 x src_len).
+        """
+        batch_size, _ = encoder_mask.shape
+        log_probs, alignment_matrix, decoder_hiddens = self.init_decoding(
+            encoder_out, encoder_mask
+        )
+        predictions, likelihood = self.greedy_step(
+            log_probs, alignment_matrix[:, 0].unsqueeze(1)
+        )
+        finished = (
+            torch.zeros(batch_size, device=self.device).bool().unsqueeze(-1)
+        )
+        for t in range(self.max_target_length):
+            symbol = predictions[:, t]
+            log_probs, alignment_matrix, decoder_hiddens = self.decode_step(
+                symbol.unsqueeze(-1),
+                decoder_hiddens,
+                encoder_out,
+                encoder_mask,
+            )
+            likelihood = likelihood + alignment_matrix.transpose(1, 2)
+            likelihood = likelihood.logsumexp(dim=-1, keepdim=True).transpose(
+                1, 2
+            )
+            pred, likelihood = self.greedy_step(log_probs, likelihood)
+            finished = finished | (pred == self.end_idx)
+            if finished.all().item():
+                break
+            # If finished decoding, pads.
+            pred = torch.where(
+                ~finished, pred, torch.tensor(self.end_idx, device=self.device)
+            )
+            predictions = torch.cat((predictions, pred), dim=-1)
+            # Updates likelihood emissions.
+            likelihood = likelihood + self._gather_at_idx(log_probs, pred)
+        return predictions, likelihood
+
+    def greedy_step(self, log_probs, likelihood):
+        # Updates current likelihood with greedy probability prediction.
+        tgt_prob = likelihood + log_probs.transpose(1, 2)
+        tgt_prob = tgt_prob.logsumexp(dim=-1)
+        tgt = torch.max(tgt_prob, dim=-1)[1]
+        return tgt.unsqueeze(-1), likelihood
+
+    def _gather_at_idx(self, prob, tgt):
```

Comment:
ditto from previous note.

Reply:
Also, A comment in the code, whether in the method signature or elsewhere, about why we need this would be really helpful!

Reply:
Oh, it's just a sophisticated form of gather since it needs to aggregate across all states for the pseudo hmm calculation. I'm kinda not a fan of it and just cribbed from original code. need to figure out a less painful one, so skimped on documentation.